src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -336,51 +261,183 @@ void parseCvrFile(List<CastVoteRecord> castVoteRecords)
-            Logger.log(Level.INFO, "Parsed %d cast vote records.", castVoteRecords.size());
+      // parse selections
+      List<Pair<Integer, String>> rankings = new ArrayList<>();
+      ArrayList cvrContestSelections = (ArrayList) cvrContest.get("CVRContestSelection");
+      for (Object cvrContestSelectionObject : cvrContestSelections) {
+        HashMap cvrContestSelection = (HashMap) cvrContestSelectionObject;
+        String contestSelectionId = (String) cvrContestSelection.get("ContestSelectionId");
+        HashMap contestSelection = (HashMap) contestSelections.get(contestSelectionId);
+        String candidateId;
+        if (contestSelection.containsKey("IsWriteIn") && contestSelection.get("IsWriteIn")
+            .equals(BOOLEAN_TRUE)) {
+          // this is a write-in
+          candidateId = this.config.getUndeclaredWriteInLabel();
+        } else {
+          // lookup Candidate Name
+          ArrayList candidateIds = (ArrayList) contestSelection.get("CandidateIds");
+          if (candidateIds == null || candidateIds.size() == 0) {
+            Logger.severe("CandidateSelection \"%s\" has no CandidateIds!", contestSelectionId);
+            throw new CvrParseException();
+          }
+          if (candidateIds.size() > 1) {
+            Logger.warning("CandidateSelection \"%s\" has multiple CandidateIds.  "
+                + "Only the first one will be processed.", contestSelectionId);
+          }
+          String candidateObjectId = (String) candidateIds.get(0);
+          HashMap Candidate = (HashMap) candidates.get(candidateObjectId);
+          candidateId = (String) Candidate.get("Name");
+          if (candidateId.equals(config.getOvervoteLabel())) {
+            candidateId = Tabulator.EXPLICIT_OVERVOTE_LABEL;
+          }
+          if (!this.config.getCandidateCodeList().contains(candidateId)) {
+            Logger.severe("Unrecognized candidate found in CVR: %s", candidateId);
+            unrecognizedCandidateCounts.merge(candidateId, 1, Integer::sum);
+          }
+        }
+
+        // parse rankings
+        // rank may appear on the CVRContestSelection OR the SelectionPosition
+        // this is an ambiguity in the nist spec
+        if (cvrContestSelection.containsKey("Rank")) {
+          Integer rank = (Integer) (cvrContestSelection.get("Rank"));
+          rankings.add(new Pair<>(rank, candidateId));
+        } else {
+          // extract all the SelectionPositions (ranks) which this selection has been assigned
+          ArrayList selectionPositions = (ArrayList) cvrContestSelection.get("SelectionPosition");
+          for (Object selectionPositionObject : selectionPositions) {
+            HashMap selectionPosition = (HashMap) selectionPositionObject;
+            // WriteIn can be linked at the selection position level
+            if (selectionPosition.containsKey("CVRWriteIn")) {
+              candidateId = this.config.getUndeclaredWriteInLabel();
```

Comment:
what if `getUndeclaredWriteInLabel()` returns null?

Reply:
@tarheel what do you think we should do?  Require that CDF tabulations have an UWI label?

Reply:
I would just throw an error here if the UWI label is null.